### PR TITLE
adds new exchange dYdX fixes #9272

### LINKED ISSUE #9272: New Exchange: dYdX
# dYdX
Add support for dYdX's Layer 2 perpetuals product to CCXT. dYdX is a hybrid decentralized exchange meaning it uses both traditional APIs while still being smart contract based.

### dYdX Developer Documentation
https://docs.dydx.exchange

### dYdX Clients
dYdX offers both TypeScript and Python clients in addition to REST and Websocket APIs. These are documented in their documentation

### dYdX Website
https://dydx.exchange/

### PR DESCRIPTION
**Pull request by tkembo**
Here's my first shot at the integration. I got stuck on authentication, specifically signing requests. 
According to the [API documentation](https://docs.dydx.exchange/?json#creating-and-signing-requests): _The header DYDX-SIGNATURE is an EIP-712-compliant Ethereum signature on a message containing the fields_
I've been going through the [CCXT Manual](https://github.com/ccxt/ccxt/wiki/Manual#authentication) and it doesn;'t look like there's support for EIP-712-compliant Ethereum signatures. 
I'm probably missing something. Can you point me in the right direction @kroitor

### CODE REVIEW

**Comment by frosty00** [comment]
> I've been going through the CCXT Manual and it doesn;'t look like there's support for EIP-712-compliant Ethereum signatures.

@tkembo look at how it is done in js/idex.js

File: js/dydx.js [hunk 1]
```diff
@@ -0,0 +1,939 @@
...
+            }
+        }
+        headers = {
+            'Content-Type': 'application/json',
+        };
+        if (this.privateKey !== undefined) {
+            headers['DYDX-SIGNATURE'] = this.privateKey;
+        }
+        if (this.walletAddress !== undefined) {
+            headers['DYDX-ETHEREUM-ADDRESS'] = this.walletAddress;
```
Comment (frosty00) [inline js/dydx.js:922]: this doesn't look right

File: js/dydx.js [hunk 2]
```diff
@@ -0,0 +1,939 @@
...
+                url = url + '?' + query;
+            } else {
+                body = this.json (params);
+            }
+        }
+        headers = {
+            'Content-Type': 'application/json',
+        };
+        if (this.privateKey !== undefined) {
+            headers['DYDX-SIGNATURE'] = this.privateKey;
```
Comment (frosty00) [inline js/dydx.js:919]: this doesn't look right

Comment (tkembo) [reply js/dydx.js]: @frosty00 I know. I wanted to first get some feedback on EIP-712-compliant Ethereum signatures before adding a signature there. Anyway, I've copied what was done on js/idex.js. 
Does this look right?

**Review by mrbaguvix — CHANGES_REQUESTED** [review]
this module is incomplete, you don't have the  `handleErrors` function. Please look at documentation.

**Comment by tkembo** [comment]
Thanks @mrbaguvix I've added a handlesErrors function but for some reason I notice that the build keeps failing. Do you know why?

**Comment by mrbaguvix** [comment]
Can you post a screenshot of your **debug/console logs** so the community can help.
CC: @kroitor or @frosty00

**Review by mrbaguvix — CHANGES_REQUESTED** [review]
Add an exception for `this.throwBroadlyMatchedException`

File: js/dydx.js [hunk 3]
```diff
@@ -0,0 +1,973 @@
...
+
+    handleErrors (code, reason, url, method, headers, body, response, requestHeaders, requestBody) {
+        if (response === undefined) {
+            return; // fallback to default error handler
+        }
+        const errorCode = this.safeString (response, 'code');
+        const message = this.safeString (response, 'error', '');
+        this.throwExactlyMatchedException (this.exceptions['exact'], message, this.id + ' ' + message);
+        this.throwExactlyMatchedException (this.exceptions['exact'], errorCode, this.id + ' ' + message);
+    }
```
Comment (mrbaguvix) [inline js/dydx.js:972]: Add an exception for `this.throwBroadlyMatchedException`

**Review by mrbaguvix — CHANGES_REQUESTED** [review]

File: js/dydx.js (hunk 3 above)
Comment (tkembo) [reply js/dydx.js:972]: image
@mrbaguvix hers' a screenshot of my debug console.

Comment (tkembo) [reply js/dydx.js]: oh, and here's the text if that's easier to read:

```
$ node run-tests --js dydx
Testing { exchanges: ["dydx"], symbol: "all", keys: { '--js': true, '--php': false, '--python': false, '--python-async': false, '--php-async': false }, maxConcurrency: 5 } (run-tests.js:255)
[100%] Testing dydx [ExchangeNotAvailable] [ExchangeNotAvailable] (testExchange @ run-tests.js:172)

WARN dydx (JavaScript): (explain @ run-tests.js:191)

	
	TESTING { exchange: "dydx", symbol: "all" }  (<anonymous> @ test.js:30)
	
	17 symbols [] (loadExchange @ test.js:183)
	SYMBOL: 1INCH-USD (testExchange @ test.js:270)
	fetching currencies not supported (exports [as fetchCurrencies] @ test.fetchCurrencies.js:37)
	1INCH-USD fetchTicker () not supported (exports [as fetchTicker] @ test.fetchTicker.js:53)
	fetching all tickers at once not supported (exports [as fetchTickers] @ test.fetchTickers.js:51)
	[ExchangeNotAvailable] dydx Not Found (tryAllProxies @ test.js:377)
	17 symbols [] (loadExchange @ test.js:183)
	SYMBOL: 1INCH-USD (testExchange @ test.js:270)
	fetching currencies not supported (exports [as fetchCurrencies] @ test.fetchCurrencies.js:37)
	1INCH-USD fetchTicker () not supported (exports [as fetchTicker] @ test.fetchTicker.js:53)
	fetching all tickers at once not supported (exports [as fetchTickers] @ test.fetchTickers.js:51)
	[ExchangeNotAvailable] dydx GET https://cors-anywhere.herokuapp.com/https://api.stage.dydx.exchange/v3/candles?market=1INCH-USD&resolution=1h&fromISO=1624024270823&limit=10 403 Forbidden See /corsdemo for more info (tryAllProxies @ test.js:377) (explain @ run-tests.js:193)
	
WARN ["dydx"] (run-tests.js:272)
All done, 1 warnings (run-tests.js:276)
```

Comment (tkembo) [reply js/dydx.js]: Thanks for pointing me in the right direction. I've now fixed is

**Review by mrbaguvix — CHANGES_REQUESTED** [review]
Thanks for your contribution @tkembo. I think for your PR to be approved, like @frosty00 had said, you should go through https://github.com/ccxt/ccxt/blob/master/js/idex.js for guide.

**Comment by tkembo** [comment]
Hi @mrbaguvix @kroitor @frosty00 - I need your help once again.

I'm trying to make a GET request but sending on of the parameters, (the market) as an [implicit method](https://github.com/ccxt/ccxt/wiki/Manual#implicit-api-methods) rather than a parameter.
I've followed [the documentation](https://github.com/ccxt/ccxt/wiki/Manual#implicit-api-methods) as well as implemented it the way it's implemented in other exchanges but I can't figure out why implementation is not working.

Here's my API definition:
```
'api': {
                'public': {
                    'get': [
                        'candles/{market}',
```

Here's my request
```
const request = {
    'market': market['id'],
};
const response = await this.publicGetCandlesMarket (this.extend (request, params));
```

But this instead of `{market}` being replaced by `1INCH-USD` for example, this is the request that's being sent out:
```
GET https://api.stage.dydx.exchange/v3/candles/{market}/?market=1INCH-USD 400 Bad Reques
```

Do you know how I can fix this?

**Comment by kroitor** [comment]
@tkembo 

There should be a call `implodeParams()` in the `sign()` method, to actually make it replace `{market}` in the url with `1INCH-USD` instead of adding it as an url-query argument. See how that is done in the other exchanges (ctrl+f `implodeParams` over all files, or over the repository on GitHub), from there you can carry how those params get substituted in the url.

**Comment by tkembo** [comment]
Thanks @kroitor - I think I've now completed the integration.

**Comment by kroitor** [comment]
@tkembo we have to add exception mappings for InsufficientFunds, InvalidOrder, and as many other exceptions as possible (see how those are done in the other exchanges). Unfortunately, dYdX's documentation lacks that information. In that case we should either ask dYdX to document those error codes or we should do the manual testing with fake orders to reverse-engineer those exceptions.

**Comment by tkembo** [comment]
Yes. I also couldn't find those from the documentation. Let me try manual testing with fake orders to reverse-engineer those while I also try to reach out to them.

**Comment by tkembo** [comment]
Hi @kroitor 
I've added all the exception mappings I could reverse engineer from manual testing with fake orders. Keen to hear what you think.

**Review by mrbaguvix — CHANGES_REQUESTED** [review]
Just curious, why have two different files?

**Comment by tkembo** [comment]
@mrbaguvix I had seen other exchanges with a second file in the `examples/js` folder and assumed it was required. I've now removed the second file.

**Comment by tkembo** [comment]
@mrbaguvix I also wanted to check: are you merging pull requests for new exchanges these days?  The reasons I'm asking is because I was looking and I notice there are pull requests for new exchanges that have been open for several months now so I figured it would be helpful for me to have realistic expectations for when a PR is likely expected to get merged after everything is done.

**Comment by tkembo** [comment]
hi @kroitor @mrbaguvix 
I just wanted to check if you have an ETA for how long it could take for this PR to get merged if everything is fine. If you can't give an exact timeline it's fine but I just wanted to check if you can give an estimate.

**Comment by CNich** [comment]
@tkembo @kroitor  @mrbaguvix  Any update on this?

**Comment by tkembo** [comment]
@CNich I haven't been getting any responses from either the dYdX or the CCXT team. The last messages I sent were like 2 months ago.

**Comment by kroitor** [comment]
@tkembo @CNich pardon for the delays, we are trying to resolve the backlog of integration PRs now, hope to get to it as soon as possible.

**Comment by CNich** [comment]
@kroitor no worries, I understand this is a huge project. I just thought I'd bring light to it since this dex looks pretty interesting.

**Comment by dyustc** [comment]
Hi, there, @tkembo @kroitor 
I am new to dydx, I have another question on how the private API is accessed. Is this done in the sign function with three levels of authentication or is it done in the [ https://github.com/dydxprotocol/v3-client](v3-client) repo independent of this pr?

**Comment by kroitor** [comment]
@dyustc hi! In CCXT the authentication is done in the `sign()` method.

**Comment by PhyrexTsai** [comment]
@tkembo @kroitor I just submit a PR to fix some issue on the current DyDx, hope you have a time take a look, here's the link https://github.com/tkembo/ccxt/pull/1

**Comment by maxholloway** [comment]
Any updates on this work?

**Comment by ttodua** [comment]
_[fixes #9272]_

**Comment by 0xcielo** [comment]
updates?

**Pull request closed without merging**